src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1857,22 +1857,28 @@ private IEnumerable<TypeInferenceInfo> InferTypeForReturnStatement(ReturnStateme
                     return;
                 }
 
-                var ancestorExpressions = returnStatement.GetAncestorsOrThis<ExpressionSyntax>();
+                var ancestor = returnStatement.AncestorsAndSelf().FirstOrDefault(e => e is ParenthesizedLambdaExpressionSyntax || e is SimpleLambdaExpressionSyntax ||
+                    e is AnonymousMethodExpressionSyntax || e is LocalFunctionStatementSyntax);
 
                 // If we're in a lambda, then use the return type of the lambda to figure out what to
                 // infer.  i.e.   Func<int,string> f = i => { return Goo(); }
-                var lambda = ancestorExpressions.FirstOrDefault(e => e.IsKind(SyntaxKind.ParenthesizedLambdaExpression, SyntaxKind.SimpleLambdaExpression));
-                if (lambda != null)
+                if (ancestor is ParenthesizedLambdaExpressionSyntax lambda)
                 {
                     types = InferTypeInLambdaExpression(lambda);
-                    isAsync = lambda is ParenthesizedLambdaExpressionSyntax && ((ParenthesizedLambdaExpressionSyntax)lambda).AsyncKeyword.Kind() != SyntaxKind.None;
+                    isAsync = lambda.AsyncKeyword.Kind() != SyntaxKind.None;
+
                     return;
+
                 }
+                else if (ancestor is SimpleLambdaExpressionSyntax simpleLambda)
+                {
+                    types = InferTypeInLambdaExpression(simpleLambda);
```

Comment:
note: SimpleLambdaExpressionSyntax can be async too, so the fact that `isAsync = ...` isn't here is a bug (it probably doesn't unwrap the return type from Task)

Reply:
you can remove this `else if` and handle both above if you check for `is LambdaExpressionSyntax` instead

Reply:
I fixed this too. Also did some minor improvements later in the method (pattern matching once more, yay!).